### ISSUE TITLE
Line numbers for primitive constants

### ISSUE THREAD

**Issue by jparise** [thriftrw/thriftrw-go#493]
These primitive constant AST types don't carry line number information (i.e. they don't implement the `nodeWithLine` interface):

```go
// ConstantBoolean is a boolean value specified in the Thrift file.
//
//   true
//   false
type ConstantBoolean bool

// ConstantInteger is an integer value specified in the Thrift file.
//
//   42
type ConstantInteger int64

// ConstantString is a string literal specified in the Thrift file.
//
//   "hello world"
type ConstantString string

// ConstantDouble is a floating point value specified in the Thrift file.
//
//   1.234
type ConstantDouble float64
```

This is/was intentional, but I wanted to ask if there might be a path to providing `lineNumber()` implementations for these types.

**Comment by abhinav** [comment]
Hey there! Unfortunately, this was less of an intentional decision and
more of an oversight.

By the time we added support for line number tracking, we had already
released 1.0 with those type definitions. We decided that losing losing
line number tracking on constant values was okay versus making a
breaking change by converting them to structs.

Today, we're still opposed to a breaking change, but there might be a
path forward. Unfortunately, not all `ast.Node` types are hashable, so
we can't just generate, say, a `map[ast.Node]Position`.

One possibility is to track line numbers for those and other hashable
types on a map on `ast.Program` on an unexported field. Something like:

```go
type Program struct {
    // ...
    
    otherLineNums map[Node]int
}
```

If a type is hashable, then we stick its location in there while
parsing. Types that are not hashable (and so, can't go there) are all
struct types, so they have their own Line fields.

We then expose a method on Program to access line information for any
node that uses the node's lineNumber method, if any, and the map if
not. Something like:

```go
func (p *Program) LineFor(n ast.Node) int {
    if n, ok := n.(nodeWithLine); ok {
        return n.lineNumber()
    }
    return p.otherLineNums[n] // zero if unknown
}
```

---

One disadvantage of this is that the added internal state may break
large, complex deep-equals between parsed `ast.Program`s and hand-
written ones.

```go
p, err := idl.Parse(...)
...
assert.Equal(t, &Program{...}, p)
// This will begin failing because Program.otherLineNums isn't set
// in the hand-written one.
```

So an alteration to the idea above is that we don't add this state to
Program, but to a new `idl.Info` type:

```go
package idl

type Info struct {
    otherLineNums map[ast.Node]int
}

func (*Info) LineFor(ast.Node) int
```

And add an alternative to `idl.Parse`. Maybe something like the
following:

```go
package idl

// Config configures the Thrift IDL parser.
type Config struct {
    // If Info is non-nil, it'll be populated with positional data
    // about the parsed nodes.
    Info *Info
}

func (c *Config) Parse(...) (*ast.Program, error)
```

(This is similar to [go/types.Config](https://golang.org/pkg/go/types/#Config).)

---

Lastly, I've talked only about line numbers for now, but to account for
column numbers eventually (ref #349, pinterest/thriftcheck#2), we can
maybe make one minor alteration above:

```go
type Position struct {
    Line int
}

type Info struct { // or Program
    // ...
    otherPositions map[ast.Node]Position
}

func (*Info) PositionFor(ast.Node) Position
```

This leaves room for adding file names and column numbers to Position
in the future without making that a problem we have to worry about
right now.

---

What do you think of the above?

**Comment by jparise** [comment]
I think that's a solid summary.

I think I would have initially pursued the `ast.Program` path, but it's good that you've identified the annoyance that could cause (e.g. `DeepEqual`) up-front. If it wasn't for that disadvantage, that would be have a good solution.

The `Config`-style approach is pretty neat. I hadn't come across that idiom before, but I can see its benefits. I don't know if there will be other "extensions" like this that would be useful to support at this (IDL parsing) layer, but it's nice that this approach offers a paved path to implementing them.

And I'm obviously `+1` on working toward column number support. 😉
